WalletWasabi.Tests/UnitTests/WabiSabi/Integration/Participant.cs

```diff
@@ -27,17 +27,48 @@ public Participant(IRPCClient rpc, WabiSabiHttpApiClient apiClient)
 		public IRPCClient Rpc { get; }
 		public WabiSabiHttpApiClient ApiClient { get; }
 
-		public async Task InitializeAsync(int numberOfCoins, CancellationToken cancellationToken)
+		private Coin? SourceCoin { get; set; }
+
+		public async Task GenerateSourceCoinAsync(CancellationToken cancellationToken)
+		{
+			var minerKey = KeyManager.GetNextReceiveKey("coinbase", out _);
+			var blockIds = await Rpc.GenerateToAddressAsync(1, minerKey.GetP2wpkhAddress(Rpc.Network)).ConfigureAwait(false);
+			var block = await Rpc.GetBlockAsync(blockIds.First()).ConfigureAwait(false);
+			SourceCoin = block.Transactions[0].Outputs.GetCoins(minerKey.P2wpkhScript).First();
+			minerKey.SetKeyState(KeyState.Used);
+		}
+
+		public async Task GenerateCoinsAsync(int numberOfCoins, CancellationToken cancellationToken)
 		{
-			var keys = KeyManager.GetKeys().Take(numberOfCoins).ToArray();
-			foreach (var key in keys)
+			var feeRate = new FeeRate(4.0m);
+			var splitTx = Transaction.Create(Rpc.Network);
+			splitTx.Inputs.Add(new TxIn(SourceCoin!.Outpoint));
+
+			var rnd = new Random();
+			var sampling = Enumerable
+				.Range(0, numberOfCoins-1)
+				.Select(_ => rnd.NextDouble())
```

Comment:
There should be a sanity check somewhere, shouldn't be?

Reply:
I will not generate them in first place.

Reply:
Doesn't `NextDouble` provide a value in `[0.0, 1.0)`? the chance that the a small amount is uneconomical given the fee rate is non-negligible (the probability that any two numbers returned in this enumerator are sufficiently close such that the effective cost of an input with the nominal value that is their difference is unecoonomical, which increases with the number of coins) but this can be easily fixed by just adding a small constant at like 57.

Actually why is it even generated as differences between random values? You can get the same distribution by just sampling the individual values exponentially.

Reply:
I don't generate small number anymore. I split the mined coin in `n` fragments and that's why I need to use the difference.

Reply:
oops, missed the `outdated` tag, sorry!